Safe initalization of libGeoIP

dridi opened the pull request:
Initially reported by @carlosabalde, there's an unsafe (and hidden) unguarded initialization in the code that can result in a segmentation fault when ran concurrently.

This pull request fixes the race by forcing the initialization on the CLI thread, and also tackles code duplication.

I have create a 4.0 branch from the current master, this work is 4.1+ only.

daghf commented:
Thanks

dridi committed:
Initialize the geoip library on the CLI thread

The initialization code in libGeoIP isn't re-entrant so instead we get
the guarantee that it is initialized exactly once.

dridi committed:
Enable warning and turn them into errors

dridi committed:
Inline init_priv into vmod_event

dridi committed:
Extract region name lookup into its own function

dridi committed:
Introduce a table of properties

For each property, two functions are created and they vary so little
that most of the code was duplicated.

gquintard commented:
Looks good to me, I just have two points:
- I don't like that ws_copy, it's possibly feasible to simply return the string as is.
- using vcc's type convesion, we should be able to get rid of the vmod_ip_ variants.

rezan commented:
> ...ws_copy, it's possibly feasible to simply return the string as is

According to this:

https://github.com/maxmind/geoip-api-c/blob/master/libGeoIP/GeoIP.h

Its safe to directly return geoip values to VCL. So right, you can skip the workspace allocation.

dridi committed:
Test an unknown IP address

dridi committed:
Drop the IP variants of the functions

The VCL compiler does that automatically now.

dridi commented:
> using vcc's type convesion, we should be able to get rid of the vmod_ip_ variants.

Done in ccf88e5 and 6375fce

> Its safe to directly return geoip values to VCL.

I have trust issues with the library, I'll double check and push a commit for review.

dridi committed:
Skip the workspace

The strings returned are perfectly safe to use and share for the whole
lifetime of the module. There's no need to waste precious work space.

dridi commented:
> Its safe to directly return geoip values to VCL. So right, you can skip the workspace allocation.

Done in 833a533

gquintard commented:
Travis is failing, but I'm assuming it's a make -j problem. The code looks good.
